### ISSUE TITLE
select fails when using contains() for non-existing values

### ISSUE THREAD

**Issue by dewittpe** [hadley/dplyr#1834]
In version 0.4.3.9001 there the `select` call can fail if called using `contains()` and the search string passed to `contains` does not exists.  Here is an example:

``` r
set.seed(1)
packageVersion("dplyr", lib.loc = .libPaths()[1])  
packageVersion("dplyr", lib.loc = "~/R-dev")      # from commit b8ad8a9

# data sets.  Note that data_fr2 does not have the *02390 columns that data_fr1
# has.
data_fr1 <- 
  data.frame(pid = 1:5, 
             pproc01222 = runif(5),
             nproc01222 = runif(5), 
             pproc01239 = runif(5),
             nproc01239 = runif(5), 
             pproc02390 = runif(5),
             nproc02390 = runif(5), 
             pproc01111 = runif(5),
             nproc01111 = runif(5))
data_fr2 <- 
  data.frame(pid = 6:10, 
             pproc01222 = runif(5),
             nproc01222 = runif(5), 
             pproc01239 = runif(5),
             nproc01239 = runif(5), 
             pproc01111 = runif(5),
             nproc01111 = runif(5))

# Use the version of dplyr on CRAN
library(dplyr, lib.loc = .libPaths()[1])
sessionInfo()$otherPkgs$dplyr$Version
# [1] "0.4.3"

bind_rows(select(data_fr1, pid, pproc01222, nproc01222, contains("02390")),
          select(data_fr2, pid, pproc01222, nproc01222, contains("02390")))

# 
# Source: local data frame [10 x 5]
# 
#      pid pproc01222 nproc01222 pproc02390 nproc02390
#    (int)      (dbl)      (dbl)      (dbl)      (dbl)
#1      1  0.2655087 0.89838968  0.9347052 0.38611409
#2      2  0.3721239 0.94467527  0.2121425 0.01339033
#3      3  0.5728534 0.66079779  0.6516738 0.38238796
#4      4  0.9082078 0.62911404  0.1255551 0.86969085
#5      5  0.2016819 0.06178627  0.2672207 0.34034900
#6      6  0.8209463 0.78935623         NA         NA
#7      7  0.6470602 0.02333120         NA         NA
#8      8  0.7829328 0.47723007         NA         NA
#9      9  0.5530363 0.73231374         NA         NA
#10    10  0.5297196 0.69273156         NA         NA

# Use the development version
detach(package:dplyr) 
library(dplyr, lib.loc = "~/R-dev")
sessionInfo()$otherPkgs$dplyr$Version
# [1] "0.4.3.9001"

bind_rows(select(data_fr1, pid, pproc01222, nproc01222, contains("02390")),
          select(data_fr2, pid, pproc01222, nproc01222, contains("02390"))) 

#    pid pproc01222 nproc01222 pproc02390 nproc02390
#1    1  0.2655087 0.89838968  0.9347052 0.38611409
#2    2  0.3721239 0.94467527  0.2121425 0.01339033
#3    3  0.5728534 0.66079779  0.6516738 0.38238796
#4    4  0.9082078 0.62911404  0.1255551 0.86969085
#5    5  0.2016819 0.06178627  0.2672207 0.34034900
#6   NA         NA         NA         NA         NA
#7   NA         NA         NA         NA         NA
#8   NA         NA         NA         NA         NA
#9   NA         NA         NA         NA         NA
#10  NA         NA         NA         NA         NA

# What if we do this work in steps?
s1 <- select(data_fr1, pid, pproc01222, nproc01222, contains("02390"))

s2 <- select(data_fr2, pid, pproc01222, nproc01222, contains("02390"))
s2
# data frame with 0 columns and 5 rows

s2 <- select(data_fr2, pid, pproc01222, nproc01222)
s2
#   pid pproc01222 nproc01222
#1   6  0.8209463  0.7893562
#2   7  0.6470602  0.0233312
#3   8  0.7829328  0.4772301
#4   9  0.5530363  0.7323137
#5  10  0.5297196  0.6927316

bind_rows(s1, s2)
#    pid pproc01222 nproc01222 pproc02390 nproc02390
#1    1  0.2655087 0.89838968  0.9347052 0.38611409
#2    2  0.3721239 0.94467527  0.2121425 0.01339033
#3    3  0.5728534 0.66079779  0.6516738 0.38238796
#4    4  0.9082078 0.62911404  0.1255551 0.86969085
#5    5  0.2016819 0.06178627  0.2672207 0.34034900
#6    6  0.8209463 0.78935623         NA         NA
#7    7  0.6470602 0.02333120         NA         NA
#8    8  0.7829328 0.47723007         NA         NA
#9    9  0.5530363 0.73231374         NA         NA
#10  10  0.5297196 0.69273156         NA         NA
```

**Comment by hadley** [comment]
Can you please provide a minimal reprex that clearly illustrates the problem? No need to include code from a previous version that does work.

**Comment by dewittpe** [comment]
```
# Load develomental version, commit b272cf2
#library(dplyr, lib.loc = "~/R-dev")

sessionInfo()$otherPkgs$dplyr$Version
# [1] "0.4.3.9001"

# select() without contains() works
dplyr::select(mtcars, mpg) %>% head(2)
#               mpg
# Mazda RX4      21
# Mazda RX4 Wag  21

# select() with contains() and a string that does occur in the column names
dplyr::select(mtcars, mpg, contains("a")) %>% head(2)
#               mpg drat am gear carb
# Mazda RX4      21  3.9  1    4    4
# Mazda RX4 Wag  21  3.9  1    4    4

# select() FAILS when contains has a string that is not in the column nams
dplyr::select(mtcars, mpg, contains("notInColNames")) %>% head(2)
# data frame with 0 columns and 2 rows
```

**Comment by hadley** [comment]
Thanks - minimal reprex is:

```R
mtcars %>% select(mpg, contains("notInColNames"))
```

**Comment by hadley** [comment]
I think this will have to wait until the next version since interactions are fairly complicated.

**Comment by skrackow** [comment]
Maybe this expands the issue: one_of() appears not to take names() when it did so in v. 0.4.3.

script:

require(dplyr)

df <- data.frame(a = 1, b = 2)
df2 <- data.frame(c = 1, b = 2)

xs <- names(select(df2, -b))

class(names(select(df2, -b)))
class(xs)

z <- select(df, -one_of("c"))
z

zz <- select(df, -one_of(xs))
zz

zzz <- select(df, -one_of(names(select(df2, -b))))
zzz

zzzz <- select(df, -one_of(names(select(df2, -c))))
zzzz

R-studio Console:

> require(dplyr)
> 
> df <- data.frame(a = 1, b = 2)
> df2 <- data.frame(c = 1, b = 2)
> 
> xs <- names(select(df2, -b))
> 
> class(names(select(df2, -b)))
[1] "character"
> class(xs)
[1] "character"
> 
> z <- select(df, -one_of("c"))
Warning message:
In one_of("c") : Unknown variables: `c`
> z
  a b
1 1 2
> 
> zz <- select(df, -one_of(xs))
Warning message:
In one_of(xs) : Unknown variables: `c`
> zz
  a b
1 1 2
> 
> zzz <- select(df, -one_of(names(select(df2, -b))))
Warning message:
In one_of(names(select(df2, -b))) : Unknown variables: `c`
> zzz
data frame with 0 columns and 1 row
> 
> zzzz <- select(df, -one_of(names(select(df2, -c))))
Warning message:
In one_of(names(select(df2, -c))) : Unknown variables: `b`
> zzzz
data frame with 0 columns and 1 row
>

**Comment by krlmlr** [comment]
I think this has been fixed in #2289. Can you please confirm?

**Comment by dewittpe** [comment]
Using the developmental version of dplyr from commit 860082f, 
```
mtcars %>% select(mpg, contains("notInColNames"))
mtcars %>% select(mpg, contains("notInColNames"), hp)
```
works as expected.
